File: Source/app/src/main/java/com/amgregoire/mangafeed/UI/Fragments/ReaderFragment.java
package com.amgregoire.mangafeed.UI.Fragments;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.AccelerateInterpolator;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.amgregoire.mangafeed.Common.MangaEnums;
import com.amgregoire.mangafeed.MangaFeed;
import com.amgregoire.mangafeed.Models.DbManga;
import com.amgregoire.mangafeed.R;
import com.amgregoire.mangafeed.UI.BackHandledFragment;
import com.amgregoire.mangafeed.UI.Mappers.IReader;
import com.amgregoire.mangafeed.UI.Presenters.ReaderPres;
import com.amgregoire.mangafeed.UI.Services.ToolbarTimerService;
import com.amgregoire.mangafeed.UI.Widgets.NoScrollViewPager;
import com.amgregoire.mangafeed.Utils.BusEvents.ReaderPageChangeEvent;
import com.amgregoire.mangafeed.Utils.MangaLogger;
import com.github.clans.fab.FloatingActionButton;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by Andy Gregoire on 3/21/2018.
 */

public class ReaderFragment extends BackHandledFragment implements IReader.ReaderMap, ToolbarTimerService.ReaderTimerListener
{
    public final static String TAG = ReaderFragment.class.getSimpleName();
    public final static String MANGA_KEY = TAG + "MANGA";
    public final static String POSITION_KEY = TAG + "POSITION";

    @BindView(R.id.noScrollViewPagerReader) NoScrollViewPager mViewPager;
    @BindView(R.id.toolbar) Toolbar mToolbar;
    @BindView(R.id.textViewReaderChapterTitle) TextView mChapterTitle;
    @BindView(R.id.textViewReaderCurrentPage) TextView mCurrentPage;
    @BindView(R.id.textViewReaderTotalPages) TextView mTotalPages;

    @BindView(R.id.topContainer) LinearLayout mToolbarContainer;

    @BindView(R.id.relativeLayoutChapterHeader) RelativeLayout mReaderHeader;
    @BindView(R.id.relativeLayoutChapterFooter) RelativeLayout mReaderFooter;

    @BindView(R.id.fabReaderNextPage) FloatingActionButton mFABNextPage;
    @BindView(R.id.fabReaderPreviousPage) FloatingActionButton mFABPrevPage;

    private IReader.ReaderPres mPresenter;

    private ToolbarTimerService mToolBarService;
    private ServiceConnection mConnection;

    public static Fragment newInstance(DbManga dbManga, int position)
    {
        Bundle lBundle = new Bundle();
        lBundle.putParcelable(MANGA_KEY, dbManga);
        lBundle.putInt(POSITION_KEY, position);

        Fragment lFragment = new ReaderFragment();
        lFragment.setArguments(lBundle);

        return lFragment;
    }

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
    {
        View lView = inflater.inflate(R.layout.fragment_reader, null);
        ButterKnife.bind(this, lView);

        mPresenter = new ReaderPres(this);
        mPresenter.init(getArguments());

        return lView;
    }

    @Override
    public void onResume()
    {
        super.onResume();
        mPresenter.onResume();

        Window w = getActivity().getWindow();
        w.setFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION, WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
        w.setFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS, WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
    }

    @Override
    public void onPause()
    {
        super.onPause();
        mPresenter.onPause();

        Window w = getActivity().getWindow();
        showToolbar();
        w.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        w.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
    }

    @Override
    public void onDestroyView()
    {
        super.onDestroyView();
        getActivity().unbindService(mConnection);
    }

    @Override
    public void initViews()
    {
        setupToolbar();

//        ViewGroup.MarginLayoutParams lToolbarParams = (ViewGroup.MarginLayoutParams) mToolbar.getLayoutParams();
//        lToolbarParams.topMargin = getStatusBarHeight();
//
//        ViewGroup.MarginLayoutParams lFooterParams = (ViewGroup.MarginLayoutParams) mReaderFooter.getLayoutParams();
//        lFooterParams.bottomMargin = getNavBarHeight();

        if (MangaFeed.Companion.getApp().getCurrentSourceType() == MangaEnums.SourceType.NOVEL)
        {
            mFABNextPage.setVisibility(View.GONE);
            mFABPrevPage.setVisibility(View.GONE);
        }

        setupViewPager();
        setupToolbarService();

        mToolbarContainer.setPadding(0, getStatusBarHeight(), 0, 0);
        mReaderFooter.setPadding(0, 0, 0, getNavBarHeight());
    }

    /***
     * This function hides the status bar, toolbar, and reader header/footers
     *
     */
    public void hideToolbar()
    {

        mToolbarContainer.animate()
                         .translationY(-mToolbarContainer.getHeight() - getStatusBarHeight())
                         .setInterpolator(new AccelerateInterpolator())
                         .start();

        mReaderFooter.animate()
                     .translationY(mReaderFooter.getHeight() + getNavBarHeight())
                     .setInterpolator(new AccelerateInterpolator())
                     .start();

        getActivity().getWindow().getDecorView()
                     .setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);

    }

    @Override
    public void hideSystemUi()
    {
        if (getActivity().getWindow().getDecorView().getSystemUiVisibility() == 0)
        {
            getActivity().getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
        }
    }

    /***
     * This function makes the status bar, toolbar, and reader header/footers visible
     *
     */
    public void showToolbar()
    {
        mToolbarContainer.animate()
                         .translationY(mToolbarContainer.getScrollY())
                         .setInterpolator(new AccelerateInterpolator())
                         .start();

        mReaderFooter.animate()
                     .translationY(-mReaderFooter.getScrollY())
                     .setInterpolator(new AccelerateInterpolator())
                     .start();

        getActivity().getWindow().getDecorView().setSystemUiVisibility(0);

    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
    {
        inflater.inflate(R.menu.menu_toolbar_reader, menu);
    }

    private void setupToolbar()
    {
        if (getActivity() != null)
        {
            mToolbar.setTitle(mPresenter.getMangaTitle());
            mToolbar.setNavigationIcon(R.drawable.nav_back_white);
            mToolbar.setNavigationOnClickListener(view -> getActivity().onBackPressed());
        }
    }

    @Override
    public void registerAdapter(PagerAdapter adapter)
    {
        if(MangaFeed.Companion.getApp().getCurrentSourceType().equals(MangaEnums.SourceType.NOVEL))
        {
            mViewPager.setPagingEnabled(true);
        }
        mViewPager.setAdapter(adapter);
        mViewPager.setOffscreenPageLimit(1);
    }

    @Override
    public void setPagerPosition(int position)
    {
        mViewPager.setCurrentItem(position);
    }

    @Override
    public void onSingleTap()
    {
        if (getActivity().getWindow().getDecorView().getSystemUiVisibility() == 0)
        {
            hideToolbar();
        }
        else
        {
            showToolbar();
        }

        startToolbarTimer();
    }

    @Override
    public void onNextChapter()
    {
        showToolbar();
        mViewPager.incrementCurrentItem();
        stopToolbarTimer();
    }

    @Override
    public void onPrevChapter()
    {
        showToolbar();
        mViewPager.decrementCurrentItem();
        stopToolbarTimer();
    }

    @Override
    public void updateToolbars(String message, String page, String total, int position)
    {
        if (position == mViewPager.getCurrentItem())
        {
            mChapterTitle.setText(message);
            mCurrentPage.setText(page);
            mTotalPages.setText(total);
        }
    }

    @Override
    public String getTagText()
    {
        return TAG;
    }

    @Override
    public boolean onBackPressed()
    {
        return true;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState)
    {
        super.onSaveInstanceState(outState);
        mPresenter.onSaveState(outState);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState)
    {
        super.onViewStateRestored(savedInstanceState);
        mPresenter.onRestoreState(savedInstanceState);
    }

    @Override
    public void stopToolbarTimer(int chPosition)
    {
        if (chPosition == mViewPager.getCurrentItem())
        {
            mToolBarService.stopTimer();
        }
    }

    @Override
    public void startToolbarTimer(int chPosition)
    {
        if (chPosition == mViewPager.getCurrentItem())
        {
            mToolBarService.startTimer();
        }
    }

    @OnClick(R.id.fabReaderPreviousPage)
    public void onFABPrevPage()
    {
        MangaFeed.Companion.getApp().rxBus().send(new ReaderPageChangeEvent(false));
        startToolbarTimer();
    }

    @OnClick(R.id.fabReaderNextPage)
    public void onFABNextPage()
    {
        MangaFeed.Companion.getApp().rxBus().send(new ReaderPageChangeEvent(true));
        startToolbarTimer();
    }

    @OnClick(R.id.fabReaderPreviousChapter)
    public void onFABPrevChapter()
    {
        onPrevChapter();
    }

    @OnClick(R.id.fabReaderNextChapter)
    public void onFABNextChapter()
    {
        onNextChapter();
    }

    @OnClick(R.id.imageViewReaderRefresh)
    public void onRefreshClicked()
    {
        MangaFeed.Companion.getApp().makeToastShort("NOT IMPLEMENTED");
//        mPresenter.refresh();
//        stopToolbarTimer();
    }

    @OnClick(R.id.imageViewReaderScreenOrientationToggle)
    public void onScreenRotateClicked()
    {
        MangaFeed.Companion.getApp().makeToastShort("NOT IMPLEMENTED");
        startToolbarTimer();
    }

    @OnClick(R.id.imageViewReaderVerticalScrollToggle)
    public void onVerticalScrollClicked()
    {
        MangaFeed.Companion.getApp().makeToastShort("NOT IMPLEMENTED");
        startToolbarTimer();
    }

    /***
     * This function retrieves the height of the android status bar.
     *
     * @return
     */
    private int getStatusBarHeight()
    {
        int lResult = 0;
        int lResourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (lResourceId > 0)
        {
            lResult = getResources().getDimensionPixelSize(lResourceId);
        }

        return lResult;
    }

    /***
     * This function retrieves the height of the android onscreen bottom navigation bar.
     *
     * @return
     */
    private int getNavBarHeight()
    {
        int lResult = 0;
        int lResourceId = getResources().getIdentifier("navigation_bar_height", "dimen", "android");
        if (lResourceId > 0)
        {
            lResult = getResources().getDimensionPixelSize(lResourceId);
        }

        return lResult;
    }

    /***
     * This function sets up the activity viewpager.
     *
     */
    private void setupViewPager()
    {
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener()
        {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels)
            {
                // Do nothing
            }

            @Override
            public void onPageSelected(int position)
            {
                mPresenter.updateCurrentPosition(position);
            }

            @Override
            public void onPageScrollStateChanged(int state)
            {
                // Do nothing
            }
        });
    }

    /***
     * This function sets up the toolbar service, that hides the toolbar, header, footer, status bar, and nav bar
     * after a set period of time after it has been shown.
     *
     */
    private void setupToolbarService()
    {
        mConnection = new ServiceConnection()
        {

            @Override
            public void onServiceConnected(ComponentName className, IBinder service)
            {
                // We've bound to ToolbarTimerService, cast the IBinder and get ToolbarTimerService instance
                ToolbarTimerService.LocalBinder binder = (ToolbarTimerService.LocalBinder) service;
                mToolBarService = binder.getService();
                mToolBarService.setToolbarListener(ReaderFragment.this);
            }

            @Override
            public void onServiceDisconnected(ComponentName aComponent)
            {
                MangaLogger.logInfo(TAG, aComponent.flattenToShortString() + " service disconnected.");
            }
        };

        mToolBarService = new ToolbarTimerService();
        mToolBarService.setToolbarListener(this);

        Intent intent = new Intent(getActivity(), ToolbarTimerService.class);
        getActivity().bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
    }

    private void startToolbarTimer()
    {
        startToolbarTimer(mViewPager.getCurrentItem());
    }

    private void stopToolbarTimer()
    {
        stopToolbarTimer(mViewPager.getCurrentItem());
    }
}

File: Source/app/src/main/java/com/amgregoire/mangafeed/v2/ui/read/ReaderFragment.kt
package com.amgregoire.mangafeed.v2.ui.read

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.view.*
import android.view.animation.AccelerateInterpolator
import androidx.appcompat.widget.PopupMenu
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.amgregoire.mangafeed.Common.MangaEnums
import com.amgregoire.mangafeed.MangaFeed
import com.amgregoire.mangafeed.R
import com.amgregoire.mangafeed.Utils.MangaLogger
import com.amgregoire.mangafeed.Utils.SharedPrefs
import com.amgregoire.mangafeed.v2.NavigationType
import com.amgregoire.mangafeed.v2.ResourceFactory
import com.amgregoire.mangafeed.v2.enums.FollowType
import com.amgregoire.mangafeed.v2.enums.ReaderSettings
import com.amgregoire.mangafeed.v2.service.Logger
import com.amgregoire.mangafeed.v2.service.ScreenUtil
import com.amgregoire.mangafeed.v2.service.ToolbarTimerService
import com.amgregoire.mangafeed.v2.ui.base.BaseFragment
import com.amgregoire.mangafeed.v2.ui.base.FragmentNavMap
import com.amgregoire.mangafeed.v2.ui.read.adapter.ChapterPagerAdapter
import kotlinx.android.synthetic.main.fragment_manga_info2.view.*
import kotlinx.android.synthetic.main.fragment_reader2.*
import kotlinx.android.synthetic.main.fragment_reader2.view.*
import kotlinx.android.synthetic.main.widget_toolbar_2.view.*

/**
 * Created by Andy Gregoire on 3/21/2018.
 */

/***
 *
 * TODO ::
 * Because fitsSystemWindows=false
 * we need to add margin/padding to top and bottom views of reader again
 *
 */
class ReaderFragment : BaseFragment(), ToolbarTimerService.ReaderTimerListener {
    private val readerViewModel: ReaderViewModel? by lazy {
        val parent = activity ?: return@lazy null
        ViewModelProviders.of(parent).get(ReaderViewModel::class.java)
    }

    private var mToolBarService: ToolbarTimerService? = null
    private val mConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            // We've bound to ToolbarTimerService, cast the IBinder and get ToolbarTimerService instance
            val binder = service as ToolbarTimerService.LocalBinder
            mToolBarService = binder.service
            mToolBarService?.setToolbarListener(this@ReaderFragment)
        }

        override fun onServiceDisconnected(aComponent: ComponentName) {
            MangaLogger.logInfo(TAG, aComponent.flattenToShortString() + " service disconnected.")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        self = inflater.inflate(R.layout.fragment_reader2, null)
        return self
    }

    override fun onResume() {
        super.onResume()
        setupToolbarService()
        showToolbar()
        enterAndExitSystemUI()
    }

    override fun onPause() {
        super.onPause()
        showToolbar()
        enterAndExitSystemUI()

        val parent = activity ?: return
        parent.unbindService(mConnection)
        parent.window.clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
    }

    override fun onStart() {
        super.onStart()

        readerViewModel?.let { readerVm ->
            val parent = activity ?: return
            readerVm.readerInfo.observe(parent, Observer { info ->
                info ?: return@Observer
                activity ?: return@Observer

                initViews()
                setupToolbar(info.manga.name)
                self.vpReader.offscreenPageLimit = 1
                if (self.vpReader.adapter == null) self.vpReader.adapter = ChapterPagerAdapter(childFragmentManager, info.dbChapters, info.manga.isFollowing, info.manga)
                if (MangaFeed.app.currentSourceType == MangaEnums.SourceType.NOVEL) self.vpReader.setPagingEnabled(true)

                val position = readerVm.getCurrentPosition()

                if (self.vpReader.currentItem == position) return@Observer

                self.vpReader.currentItem = position

            })

            readerVm.chapterInfo.observe(parent, Observer { info ->
                info ?: return@Observer
                activity ?: return@Observer

                self.tvChapterTitle.text = info.title
                self.tvCurrentPage.text = info.currentPage.inc().toString()
                self.tvPageCount.text = info.totalPages.toString()
            })

            readerVm.uiState.observe(parent, Observer { state ->
                state ?: return@Observer

                when (state) {
                    is ReaderUIState.INIT -> showToolbar()
                    is ReaderUIState.SHOW -> {
                        showToolbar()
                        mToolBarService?.startTimer()
                    }
                    else -> {
                        Logger.error("Toggling UI State = $state")
                        hideToolbar()
                    }
                }
            })
        }

        setupListeners()
    }

    private fun setupListeners() {
        fabNextPage.setOnClickListener {
            val chapter = readerViewModel?.getChapterByPosition(self.vpReader.currentItem) ?: return@setOnClickListener
            readerViewModel?.incrementPage(chapter)
            mToolBarService?.restartTimer()
        }

        fabPrevPage.setOnClickListener {
            val chapter = readerViewModel?.getChapterByPosition(self.vpReader.currentItem) ?: return@setOnClickListener
            readerViewModel?.decrementPage(chapter)
            mToolBarService?.restartTimer()
        }

        fabNextChapter.setOnClickListener {
            readerViewModel?.incrementChapter()
            mToolBarService?.restartTimer()
        }

        fabPrevChapter.setOnClickListener {
            readerViewModel?.decrementChapter()
            mToolBarService?.restartTimer()
        }

        imageViewReaderRefresh
        imageViewReaderRefresh.setOnClickListener {
            (self.vpReader.adapter as? ChapterPagerAdapter)?.updateItemAt(self.vpReader.currentItem)
        }

        self.ivReaderSetting.setOnClickListener {
            val context = context ?: return@setOnClickListener
            val popMenu = PopupMenu(context, self.ivReaderSetting, Gravity.RIGHT)

            popMenu.menuInflater.inflate(R.menu.menu_reader_setting, popMenu.menu)

            popMenu.setOnMenuItemClickListener { popupItem ->
                when (popupItem.itemId) {
                    R.id.menuReaderSettingLow -> readerViewModel?.updateReaderSettings(ReaderSettings.Low)
                    R.id.menuReaderSettingMedium -> readerViewModel?.updateReaderSettings(ReaderSettings.Medium)
                    R.id.menuReaderSettingMax -> readerViewModel?.updateReaderSettings(ReaderSettings.Max)
                    else -> readerViewModel?.updateReaderSettings(ReaderSettings.Max)
                }
                true
            }

            popMenu.show()
        }
    }

    fun initViews() {
        if (MangaFeed.app.currentSourceType == MangaEnums.SourceType.NOVEL) {
            self.fabNextPage.visibility = View.GONE
            self.fabPrevPage.visibility = View.GONE
        }

        setupViewPager()
    }

    fun hideSystemUi() {
        val parent = activity ?: return
        parent.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                View.SYSTEM_UI_FLAG_FULLSCREEN or
                View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY

        if (SharedPrefs.isLightTheme()) parent.window.decorView.systemUiVisibility = parent.window.decorView.systemUiVisibility or
                View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR or
                View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
    }

    override fun hideSystemUiService() {
        activity ?: return
        readerViewModel?.setUIStateHide()
        hideSystemUi()
    }

    private fun showSystemUi() {
        val parent = activity ?: return
        parent.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or
                View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION

        if (SharedPrefs.isLightTheme()) parent.window.decorView.systemUiVisibility = parent.window.decorView.systemUiVisibility or
                View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR or
                View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
    }

    private fun enterAndExitSystemUI() {
        val parent = activity ?: return
        parent.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY

        if (SharedPrefs.isLightTheme()) parent.window.decorView.systemUiVisibility = parent.window.decorView.systemUiVisibility or
                View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR or
                View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR

        parent.window.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)

        self.bottomContainer.setPadding(0, 0, 0, ScreenUtil.getNavigationBarHeight(resources))
        self.topContainer.setPadding(0, ScreenUtil.getStatusBarHeight(resources), 0, 0)
    }

    /***********************************************************
     *
     * Reader Interactions
     *
     **********************************************************/
    fun onRefreshClicked() {
        MangaFeed.app.makeToastShort("NOT IMPLEMENTED")
        mToolBarService?.stopTimer()
    }

    fun onScreenRotateClicked() {
        MangaFeed.app.makeToastShort("NOT IMPLEMENTED")
        mToolBarService?.startTimer()
    }

    fun onVerticalScrollClicked() {
        MangaFeed.app.makeToastShort("NOT IMPLEMENTED")
        mToolBarService?.startTimer()
    }

    /***
     * This function sets up the activity viewpager.
     *
     */
    private fun setupViewPager() {
        self.vpReader.addOnPageChangeListener(object : androidx.viewpager.widget.ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) = Unit
            override fun onPageScrollStateChanged(state: Int) = Unit
            override fun onPageSelected(position: Int) {
                Logger.error("Chapter selected: $position")
                //                showToolbar()
                //                mToolBarService?.startTimer()
                readerViewModel?.updateCurrentChapterByPosition(position)
            }
        })
    }

    /***
     * This function sets up the toolbar service, that hides the toolbar, header, footer, status bar, and nav bar
     * after a set period of time after it has been shown.
     *
     */
    private fun setupToolbarService() {
        mToolBarService = ToolbarTimerService()
        mToolBarService?.setToolbarListener(this)

        val parent = activity ?: return
        val intent = Intent(parent, ToolbarTimerService::class.java)
        parent.bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
    }

    /***
     * This function makes the status bar, toolbar, and reader header/footers visible
     *
     */
    private fun showToolbar() {
        self.topContainer.animate()
                .translationY(self.topContainer.scrollY.toFloat())
                .setInterpolator(AccelerateInterpolator())
                .start()

        self.bottomContainer.animate()
                .translationY((-self.bottomContainer.scrollY).toFloat())
                .setInterpolator(AccelerateInterpolator())
                .start()

        showSystemUi()
    }

    /***
     * This function hides the status bar, toolbar, and reader header/footers
     *
     */
    fun hideToolbar() {
        val parent = activity ?: return
        self.topContainer.animate()
                .translationY((-self.topContainer.height - ScreenUtil.getStatusBarHeight(parent.resources)).toFloat())
                .setInterpolator(AccelerateInterpolator())
                .start()

        self.bottomContainer.animate()
                .translationY((self.bottomContainer.height + ScreenUtil.getNavigationBarHeight(parent.resources)).toFloat())
                .setInterpolator(AccelerateInterpolator())
                .start()

        hideSystemUi()
    }

    override fun hideToolbarService() {
        activity ?: return
        readerViewModel?.setUIStateHide()
        hideToolbar()
    }

    private fun setupToolbar(title: String) {
        self.toolbar.title = title
        self.toolbar.setNavigationIcon(ResourceFactory.getNavigationIcon(NavigationType.Back))
        self.toolbar.setNavigationOnClickListener {
            val parent = activity ?: return@setNavigationOnClickListener
            (parent as FragmentNavMap).popBackStack()
        }
    }

    companion object {
        val TAG: String = ReaderFragment::class.java.simpleName
        fun newInstance() = ReaderFragment()
    }
}
